### PR TITLE
fix: update timex/tzdata dependencies

### PR DESCRIPTION
**Pull request by bitwalker**
I noticed that this library is pinned to a specific range of Timex releases, but those releases have various bugs that have been addressed for some time now, and by pinning the range, it constrains downstream consumers from being able to upgrade without explicitly overriding the version of Timex used (and thus likely causing things to break if assumptions are being made that rely on behavior in a specific version). This PR changes the constraint to require at least 3.7. There are no API changes that will occur in further versions of the `3.x` release series, so this shouldn't present an issue with future releases.

### CODE REVIEW

**Comment by bitwalker** [comment]
@ananthakumaran I think I'll need your help in understanding what's going on with the failing tests here. I started by looking at `exq_scheduler_time_test.exs` and `daylight_saving_test.exs`, but I get weird behavior running some of those tests locally, and I'm not clear on how those suites are set up and what they are intended to test to figure it out. 

The other tests failures seem to be due to a chain reaction which causes the Redix client to crash in all subsequent tests, but I think that's unrelated to the actual changes and just a property of how the test suite runs currently. That said, there may be other failures hiding once I fix those in the two suites above, I just want to make sure I know what I'm doing in there.

**Comment by ananthakumaran** [comment]
I think the redix failures are mostly unrelated (some of the tests kill the connection via toxiproxy, which could lead to a lot of logs). I haven't spent much time on debugging, but a quick look at the value returned by the old and new implementation of `local_to_utc(naive_time, timezone)` doesn't seem to be the same

for naive_time `~N[2019-03-31 02:20:00]` and timezone `Europe/Copenhagen`, the old implementation returns `~U[2019-03-31 01:00:00Z]` whereas the new implementation returns `#<Ambiguous(~U[2019-03-31 00:59:59.999999Z] ~ ~U[2019-03-31 01:00:00Z])>`. 

To rule out tzinfo version change, I bumped the version in master branch and ran the tests, it passes without any failures.

So the failures seem to be from the differences in Timex version?

**Comment by bitwalker** [comment]
@ananthakumaran The new result is actually correct, it just looks weird because of the conversion to UTC of both the before/after date/times, but because that edge case is trivially resolved to a single DateTime, I've modified that function a bit to handle it.

The only tests that I'm not sure about are in `test/exq_scheduler_time_test.exs` - there is some flakiness with at least one of them, which sometimes fails because jobs haven't been queued yet (I've added an assertion to that test to raise an error when that happens). The other two I'm just not familiar enough with the test setup to understand why they are failing. I'm hoping you will have some ideas on where I can look, or even what the actual problem is.

A couple other notes:

- I noticed there is a race condition with the fake `Time` module that gets compiled during tests, so I've moved things around a bit to ensure that support modules are compiled in advance.
- There is the issue of supporting Elixir 1.6 moving forward, as `System.stacktrace/0` was removed from new versions of Elixir, while `__STACKTRACE__` is not present on 1.6. I think it is probably time to release 1.0 and remove support for 1.6 moving forward (if not 1.8 as well). Those still on 1.6 can continue to use the 0.x releases, but the project will be able to keep up with newer releases (as well as updates to its deps)

**Comment by ananthakumaran** [comment]
> I noticed there is a race condition with the fake Time module that gets compiled during tests, so I've moved things around a bit to ensure that support modules are compiled in advance.

I cherry-picked your second commit, so things would be easier to debug (please rebase from master again)

> The only tests that I'm not sure about are in test/exq_scheduler_time_test.exs - there is some flakiness with at least one of them, which sometimes fails because jobs haven't been queued yet (I've added an assertion to that test to raise an error when that happens). The other two I'm just not familiar enough with the test setup to understand why they are failing. I'm hoping you will have some ideas on where I can look, or even what the actual problem is.

I will try to debug it, the code is written by multiple persons some time ago and my mental model is not very good either. I will let you know if I discover anything.

> There is the issue of supporting Elixir 1.6 moving forward, as System.stacktrace/0 was removed from new versions of Elixir, while __STACKTRACE__ is not present on 1.6. I think it is probably time to release 1.0 and remove support for 1.6 moving forward (if not 1.8 as well). Those still on 1.6 can continue to use the 0.x releases, but the project will be able to keep up with newer releases (as well as updates to its deps)

agree, 1.8+ might be okay now. but let's do that on a separate PR. In this one, let's just focus on fixing timex version only.

**Comment by ananthakumaran** [comment]
I think, I found the issue. The following patch fixes the test failure
```diff
--- a/lib/exq_scheduler/schedule.ex
+++ b/lib/exq_scheduler/schedule.ex
@@ -133,6 +133,7 @@ defmodule ExqScheduler.Schedule do
         schedule_last_run_time =
           schedule_last_run_time
           |> Timex.parse!("{ISO:Extended:Z}")
+          |> Timex.to_datetime("Etc/UTC")

         Utils.get_nearer_date(ref_time, lower_bound_time, schedule_last_run_time)
```

As to why, UTC time gets stored in Redis and later parsed back as naive instead of UTC. Now, this should have cause failure earlier, but this was compensated by the way `utc_to_localtime` function works in master. Especially, the older Timezone.convert would treat the naive date as UTC and convert to the target zone. The newer version doesn't treat the naive date as UTC which exposes the initial bug.

```
master
iex(2)> x = "2021-08-05T15:07:16.111769"
"2021-08-05T15:07:16.111769"
iex(3)> p = Timex.parse!(x, "{ISO:Extended:Z}")
~N[2021-08-05 15:07:16.111769]
iex(5)> Timex.Timezone.convert(p, "Asia/Kolkata")
#DateTime<2021-08-05 20:37:16.111769+05:30 IST Asia/Kolkata>
iex(4)> Timex.Timezone.convert(p, "Asia/Kolkata") |> DateTime.to_naive()
~N[2021-08-05 20:37:16.111769]

pr
iex(1)> x = "2021-08-05T15:07:16.111769"
"2021-08-05T15:07:16.111769"
iex(3)> p = Timex.parse!(x, "{ISO:Extended:Z}")
~N[2021-08-05 15:07:16.111769]
iex(6)> Timex.Timezone.convert(p, "Asia/Kolkata")
#DateTime<2021-08-05 15:07:16.111769+05:30 IST Asia/Kolkata>
iex(7)> Timex.Timezone.convert(p, "Asia/Kolkata") |> DateTime.to_naive()
~N[2021-08-05 15:07:16.111769]
```

**Comment by ananthakumaran** [comment]
I just realized Timex itself seems to depend on __STACKTRACE__, in that are we can drop 1.6 support and mark the library as 1.7+? feel free to adjust the matrix

File: test/exq_scheduler_time_test.exs [hunk 1]
```diff
@@ -45,6 +45,8 @@ defmodule ExqSchedulerTimeTest do
 
     jobs = get_jobs("TimeWorker", "TimeQ")
 
+    assert length(jobs) > 0
```
Comment (ananthakumaran) [inline test/exq_scheduler_time_test.exs:48]: any specific reason for this assert? as this fails locally for me as well.

Comment (bitwalker) [reply test/exq_scheduler_time_test.exs:48]: I added it because the following assertion will pass even if there are no jobs, which seems like it would be misleading at best

Comment (bitwalker) [reply test/exq_scheduler_time_test.exs:48]: I think that this test has been broken for quite some time, I can replicate the same failure on master just by adding this assertion

Comment (bitwalker) [reply test/exq_scheduler_time_test.exs:48]: It looks like that test is the only thing failing the build at this point, so I'm going to leave the assertion in, but tag the test as `skip: true`

Comment (ananthakumaran) [reply test/exq_scheduler_time_test.exs:48]: I have adjusted the sleep time to fix this issue
